rush-plugins/rush-operation-resource-plugin/src/RushOperationResourcePlugin.ts

```diff
@@ -0,0 +1,214 @@
+  private readonly _options: IRushOperationResourcePluginOptions;
+  private readonly _checkedOutResources: Map<Operation, string>;
+  private readonly _resourcePools: Map<string, IResourcePool>;
+
+  public constructor(options: IRushOperationResourcePluginOptions) {
+    this._options = options;
+    this._checkedOutResources = new Map();
+    this._resourcePools = new Map();
+  }
+
+  public apply(rushSession: RushSession, rushContext: RushConfiguration): void {
+    const logger: ILogger = rushSession.getLogger(this.pluginName);
+
+    this._validateOptions();
+
+    rushSession.hooks.runAnyPhasedCommand.tapPromise(
+      this.pluginName,
+      async (command: IPhasedCommand): Promise<void> => {
+        command.hooks.beforeExecuteOperation.tapPromise(
+          this.pluginName,
+          async (
+            runnerContext: IOperationRunnerContext & IOperationExecutionResult
+          ): Promise<OperationStatus | undefined> => {
+            const constraint: IResourceConstraintConfig | undefined = this._getConstraintsForOperation(
+              runnerContext.operation
+            );
+
+            if (constraint) {
+              if (constraint.resourcePool) {
+                const pool: IResourcePool = this._resourcePools.get(constraint.resourcePool.poolName)!;
+
+                // Wait to obtain a resource from the resource pool for this constraint
+                const resource: string = await this._obtainResource(pool, runnerContext.operation);
+
+                // Pass this resource to the underlying operation using the specified
+                // environment variable
+                if (constraint.resourcePool.envVarName) {
+                  const runner: any = runnerContext.operation.runner;
+                  if (!runner.environment) {
+                    runner.environment = {};
+                  }
+                  runner.environment[constraint.resourcePool.envVarName] = resource;
+                }
+
+                // Reset the stopwatch start time to account for the time spent waiting
+                (runnerContext.stopwatch as Stopwatch).reset().start();
+
+                // In "rush --debug" mode, log every time we assign a resource
+                logger.terminal.writeVerboseLine(
+                  `[${this.pluginName}] ${runnerContext.operation.associatedProject?.packageName} (${runnerContext.operation.associatedPhase?.name}) Assigned ${constraint.resourcePool.envVarName}=${resource} from pool ${pool.poolName}.`
+                );
+              }
+            }
+
+            return;
+          }
+        );
+
+        command.hooks.afterExecuteOperation.tapPromise(
+          this.pluginName,
+          async (runnerContext: IOperationRunnerContext & IOperationExecutionResult): Promise<void> => {
+            const constraint: IResourceConstraintConfig | undefined = this._getConstraintsForOperation(
+              runnerContext.operation
+            );
+
+            if (constraint) {
```

Comment:
I suggest having your resource assignments implement the `[Symbol.disposable](): void` contract and just store a `WeakMap<Operation, IDisposable[]>` of the resources that have been assigned to each operation, then in this step, iterate and dispose all resources, then delete the map entry.